felix/fv/bpf_programs_test.go

```diff
@@ -92,4 +94,48 @@ var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ Felix bpf reattach object",
 		Eventually(secondRunProg2, "10s", "100ms").Should(BeClosed())
 		Expect(secondRunBase).NotTo(BeClosed())
 	})
+
+	It("should not leak bpf programs", func() {
+		noOfBPFPrograms := func() int {
+			out, err := felixes[0].ExecCombinedOutput("bpftool", "prog", "list")
+			Expect(err).NotTo(HaveOccurred())
+			noOfProgs := strings.Count(out, "sched_cls")
+			return noOfProgs
+		}
+
+		const noOfWorkloads = 5
+		var wl [noOfWorkloads]*workload.Workload
+		initVal := noOfBPFPrograms()
+		expectedWorkloadPrograms := 2*5 + 2 // 5 programs for each hook + 2 policy programs
+		expectedHostPrograms := 2 * 5       // 5 programs for each hook
+
+		felixes[0].TriggerDelayedStart()
+		Eventually(noOfBPFPrograms, "10s", "100ms").Should(Equal(initVal + expectedHostPrograms))
+
+		By("creating workloads")
+		for i := 0; i < noOfWorkloads; i++ {
+			wl[i] = workload.New(felixes[0], fmt.Sprintf("w%d", i), "default", fmt.Sprintf("10.65.0.%d", i+2), "8085", "tcp")
+			err := wl[i].Start()
+			Expect(err).ToNot(HaveOccurred())
+			wl[i].ConfigureInInfra(infra)
+			Eventually(noOfBPFPrograms, "10s", "100ms").Should(Equal(initVal + (i+1)*expectedWorkloadPrograms + expectedHostPrograms))
```

Comment:
I wonder if all these `Eventually` just make the tests run nicely. I think we would need more of a fuzzy test that randomly in parallel creates and tears down a bunch of workloads and then after a while checks that the numbers match. There can be some race somewhere and a nice sync test may not trigger that.

We may avoid running the test regularly or we can configure the test so that it normally runs for a short period only and then use it as a base for testing when we need it :thinking:

Reply:
I ran the test more about 50 times, and it looks that it works fine. There should not be any flake.

But about fuzzing, do you think FV is the right place? golang 1.18 seems to have a fuzzer, maybe we can use it. Or do you any other tool?